### ISSUE TITLE
Odd token length trigger a "Odd-length string" server error

### ISSUE THREAD

**Issue by metathink** [James1345/django-rest-knox#46]
Version: django-rest-knox==2.3.0

**Produces an error 401:**
- Authorization: Token 12
- Authorization: 123

**Produces an "Odd-length string" error 500:**
- Authorization: Token 123
- Authorization: Token abcde
- ... every other token with an odd length

**Comment by belugame** [comment]
@metathink I have seen this behavior before and we introduced a testcase for it though it was hard to replicate and seeing your report now I'm not 100% convinced the testcase does do that. Can you test with Version 3.0?

**Comment by metathink** [comment]
Same behavior with django-rest-knox==3.0.0

The end of my traceback:
```
  File "/abc/venv/lib/python3.6/site-packages/knox/auth.py", line 45, in authenticate
    user, auth_token = self.authenticate_credentials(auth[1])
  File "/abc/venv/lib/python3.6/site-packages/knox/auth.py", line 67, in authenticate_credentials
    digest = hash_token(token, auth_token.salt)
  File "/abc/venv/lib/python3.6/site-packages/knox/crypto.py", line 29, in hash_token
    digest.update(binascii.unhexlify(token))
binascii.Error: Odd-length string
```

**Comment by belugame** [comment]
@metathink I cannot reproduce this with either version. I'm testing it with httpie:

```
http -vj my.host/api/v1/ "Authorization:Token 123"

GET /api/v1/ HTTP/1.1
Accept: application/json, */*
Accept-Encoding: gzip, deflate
Authorization: Token 123
Connection: keep-alive
Content-Type: application/json
Host: my.host
User-Agent: HTTPie/0.9.6



HTTP/1.1 401 UNAUTHORIZED
Allow: GET, HEAD, OPTIONS
Connection: Keep-Alive
Content-Type: application/json; charset=utf-8
Date: Mon, 13 Mar 2017 18:18:48 GMT
Keep-Alive: timeout=5, max=100
Set-Cookie: sessionid=xxx; expires=Mon, 13-Mar-2017 22:18:48 GMT; httponly; Max-Age=14400; Path=/
Transfer-Encoding: chunked
Vary: Accept,Accept-Language,Cookie
WWW-Authenticate: Token

{
    "detail": "Ungültiges Token"
}
```

It replies with a 401 status and message "Invalid token" which is as expected.
Please provide a failing testcase if possible.

**Comment by belugame** [comment]
Please reopen if needed with more info.

**Comment by chdsbd** [comment]
I ran into this when I passed a token longer than the default token.

**Comment by belugame** [comment]
@chdsbd what HTTP status integer do you receive as a response?

**Comment by chdsbd** [comment]
I receive a 500 error.

I simple add a few chars to the end of `13a40aa6751cc6e4e8005a7a3cc39301a0d856b72296a3b8a6c0d23f7c676670` and have this string `13a40aa6751cc6e4e8005a7a3cc39301a0d856b72296a3b8a6c0d23f7c676670-1234` which will cause the error.

Here is the stack trace:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 41, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/base.py", line 187, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/base.py", line 185, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.6/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/django/views/generic/base.py", line 68, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 489, in dispatch
    response = self.handle_exception(exc)
  File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 449, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 477, in dispatch
    self.initial(request, *args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 394, in initial
    self.perform_authentication(request)
  File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 320, in perform_authentication
    request.user
  File "/usr/local/lib/python3.6/site-packages/rest_framework/request.py", line 378, in __getattribute__
    return super(Request, self).__getattribute__(attr)
  File "/usr/local/lib/python3.6/site-packages/rest_framework/request.py", line 196, in user
    self._authenticate()
  File "/usr/local/lib/python3.6/site-packages/rest_framework/request.py", line 342, in _authenticate
    user_auth_tuple = authenticator.authenticate(self)
  File "/usr/local/lib/python3.6/site-packages/knox/auth.py", line 51, in authenticate
    user, auth_token = self.authenticate_credentials(auth[1])
  File "/usr/local/lib/python3.6/site-packages/knox/auth.py", line 74, in authenticate_credentials
    digest = hash_token(token, auth_token.salt)
  File "/usr/local/lib/python3.6/site-packages/knox/crypto.py", line 29, in hash_token
    digest.update(binascii.unhexlify(token))
binascii.Error: Odd-length string
```

**Comment by belugame** [comment]
@chdsbd  Ok, just reopened the issue. If you could provide a failing test case that would be most helpful.

**Comment by belugame** [comment]
fixed in release 3.1.3. Thanks to @chdsbd!
